server/src/modules/maps/maps.service.ts

```diff
@@ -440,6 +449,226 @@ export class MapsService {
+        if (query.zoneNum) where.run.zoneNum = query.zoneNum;
+
+        const include = {
+            run: true
+        };
+
+        const dbResponse = await this.mapRepo.getRank(where, include);
+
+        if (!dbResponse) throw new NotFoundException('Rank not found');
+
+        //This is done because the MapRankDto still contains trackNum and zoneNum but UserMapRank doesn't
+        return DtoFactory(MapRankDto, {
+            mapID: dbResponse.mapID,
+            userID: dbResponse.userID,
+            runID: dbResponse.runID,
+            gameType: dbResponse.gameType,
+            flags: dbResponse.flags,
+            trackNum: (dbResponse as any).run.trackNum,
+            zoneNum: (dbResponse as any).run.zoneNum,
+            rank: dbResponse.rank,
+            rankXP: dbResponse.rankXP,
+            createdAt: dbResponse.createdAt,
+            updatedAt: dbResponse.updatedAt
+        });
+    }
+
+    async getRankAround(userID: number, mapID: number, query: MapRankGetNumberQuery): Promise<MapRankDto[]> {
+        const where: Prisma.UserMapRankWhereInput = {
+            mapID: mapID,
+            flags: 0,
+            userID: userID,
+            run: {
+                trackNum: 0,
+                zoneNum: 0
+            }
+        };
+
+        if (query.flags) where.flags = query.flags;
+        if (query.trackNum) where.run.trackNum = query.trackNum;
+        if (query.zoneNum) where.run.zoneNum = query.zoneNum;
+
+        const include = {
+            run: true
+        };
+
+        const order: Prisma.UserMapRankOrderByWithAggregationInput = {
+            rank: 'asc'
+        };
+
+        const myRankInfo = await this.mapRepo.getRank(where, include);
+
+        if (!myRankInfo) throw new NotFoundException('No personal best detected');
+
+        const myRank = myRankInfo.rank;
+
+        where.userID = undefined;
+
+        const ranks = await this.mapRepo.getRanks(where, include, order, Math.max(myRank - 6, 0), 11);
+
+        //This is done because the MapRankDto still contains trackNum and zoneNum but UserMapRank doesn't
+        const ret = ranks[0].map((item) =>
+            DtoFactory(MapRankDto, {
+                mapID: item.mapID,
+                userID: item.userID,
+                runID: item.runID,
+                gameType: item.gameType,
+                flags: item.flags,
+                trackNum: (item as any).run.trackNum,
+                zoneNum: (item as any).run.zoneNum,
+                rank: item.rank,
+                rankXP: item.rankXP,
+                createdAt: item.createdAt,
+                updatedAt: item.updatedAt
+            })
+        );
+
+        return ret;
+    }
+
+    async getRankFriends(userID: number, mapID: number, query: MapRankGetNumberQuery): Promise<MapRankDto[]> {
+        const user = await this.userRepo.get(userID);
+        const steamID = user.steamID;
+
+        const res = await firstValueFrom(
```

Comment:
Query is mostly right but I think we should split this up into its own function in `users.service.ts`. You'll need to update `maps.module.ts` to import the `UsersService`, I don't mind doing that.

Then let's type up the response from Steam, you can add that to `auth.interface.ts` for now. Maybe not the best place for that file, in the future we can move that out to something like `lib/steam/interfaces.ts`.

Reply:
Make sure you place this in a good location in the users.service file. Change the "Private" region to "Steam Queries" and add there (public of course, leave rest private until they're needed elsewhere)